cpp/src/join/semi_join.cu

```diff
@@ -136,20 +167,47 @@ std::unique_ptr<rmm::device_uvector<cudf::size_type>> left_semi_anti_join(
   auto gather_map =
     std::make_unique<rmm::device_uvector<cudf::size_type>>(left_num_rows, stream, mr);
 
-  // gather_map_end will be the end of valid data in gather_map
-  auto gather_map_end = thrust::copy_if(
+  // Copy selected rows from the hash table
+  rmm::device_uvector<size_type> num_selected(1, stream, mr);
+  rmm::device_uvector<bool> flagged(left_num_rows, stream, mr);
+  auto flagged_d = flagged.data();
+
+  // Create boolean array flagging which rows to copy
+  auto counting_iter = thrust::counting_iterator<size_type>(0);
+  thrust::for_each(
     rmm::exec_policy(stream),
-    thrust::make_counting_iterator(0),
-    thrust::make_counting_iterator(left_num_rows),
-    gather_map->begin(),
-    [hash_table_view, join_type_boolean, hash_probe, equality_probe] __device__(
-      size_type const idx) {
-      // Look up this row. The hash function used here needs to map a (left) row index to the hash
-      // of the row, so it's a row hash. The equality check needs to verify
-      return hash_table_view.contains(idx, hash_probe, equality_probe) == join_type_boolean;
+    counting_iter,
+    counting_iter + left_num_rows,
+    [flagged_d, hash_table_view, join_type_boolean, hash_probe, equality_probe] __device__(
+      const size_type idx) {
+      flagged_d[idx] =
+        hash_table_view.contains(idx, hash_probe, equality_probe) == join_type_boolean;
     });
 
-  auto join_size = thrust::distance(gather_map->begin(), gather_map_end);
+  auto copy_rows = [&](void* temp_storage, size_t& temp_bytes) {
+    cub::DeviceSelect::Flagged(temp_storage,
+                               temp_bytes,
+                               cub::CountingInputIterator<size_type>(0),
+                               flagged_d,
+                               gather_map->begin(),
+                               num_selected.data(),
+                               left_num_rows,
+                               stream);
+  };
+
+  // Get how much temp space we need
+  size_t flagged_temp_bytes = 0;
+  copy_rows(nullptr, flagged_temp_bytes);
+
+  rmm::device_uvector<unsigned char> flagged_tmp(flagged_temp_bytes, stream, mr);
```

Comment:
```suggestion
  rmm::device_buffer flagged_tmp(flagged_temp_bytes, stream, mr);
```

Reply:
No need to type the data if the function expects it to be `void*` anyways.

Reply:
No longer used